In [1]:
import pandas as pd
from sqlalchemy import create_engine
!pip install pymysql sqlalchemy

**📝 Explanation:** This cell performs the following task:

>> _[Importing necessary libraries]_

In [2]:
import pandas as pd
from sqlalchemy import create_engine

import pandas as pd
df_agg_user_full = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\aggregated_user_etl.csv")
df_agg_transaction = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\aggregated_transaction_etl.csv")
df_agg_insurance = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\aggregated_insurance_etl.csv")
df_map_user = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\map_user_etl.csv")
df_map_map = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\map_map_etl.csv")
df_map_insurance = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\map_insurance_etl_fixed.csv")
df_top_user = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\top_user_etl.csv")
df_top_map = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\top_map_etl.csv")
df_top_insurance = pd.read_csv("C:\\Users\\vikas\\OneDrive\\Desktop\\LMS 6th Project PhonePe\\data\\top_insurance_etl.csv")

username = "root"
password = "vikasashi"
host = "localhost"
port = "3306"
database = "phonepe1"

# Creating SQLAlchemy engine for MySQL
mysql_url = f"mysql+pymysql://{username}:{password}@{host}:{port}/{database}"
engine = create_engine(mysql_url)

# Loading DataFrame into the tables
df_agg_user_full.to_sql("aggregated_user", con=engine, if_exists="append", index=False)
df_agg_transaction.to_sql("aggregated_transaction", con=engine, if_exists="append", index=False)
df_agg_insurance.to_sql("aggregated_insurance", con=engine, if_exists="append", index=False)
df_map_user.to_sql("map_user", con=engine, if_exists="append", index=False)
df_map_map.to_sql("map_map", con=engine, if_exists="append", index=False)
df_map_insurance.to_sql("map_insurance", con=engine, if_exists="append", index=False)
df_top_user.to_sql("top_user", con=engine, if_exists="append", index=False)
df_top_map.to_sql("top_map", con=engine, if_exists="append", index=False)
df_top_insurance.to_sql("top_insurance", con=engine, if_exists="append", index=False)

12276

**📝 Explanation:** This cell performs the following task:

> _[In this section, we have used sqlalchemy to perform sql queries directly into the python to save code and time]_
>
> The dataset was given in json format, I have converted it into csv files to handle dataset easily.
> Above code imports the dataset through SQL query into SQL Database

In [3]:
query_results = {}

In [4]:
sql_queries = {
    "Device Brand Penetration (Top Brands)": """
        SELECT brand, SUM(brand_count) AS user_count
        FROM aggregated_user
        GROUP BY brand
        ORDER BY user_count DESC
        LIMIT 10;
    """,

    "Top States by Transaction Consistency (Low Std Dev)": """
        SELECT state, STDDEV_SAMP(transaction_amount) AS std_dev
        FROM aggregated_transaction
        GROUP BY state
        HAVING COUNT(*) > 1
        ORDER BY std_dev ASC
        LIMIT 10;
    """,

    "Fastest Growing Insurance Districts (Q-over-Q Growth)": """
        SELECT district, year, quarter, SUM(amount) AS total_insurance
        FROM map_insurance
        GROUP BY district, year, quarter
        ORDER BY district, year, quarter;
    """,


    "Top Pin Codes by Transaction Value": """
        SELECT pincode, SUM(amount) AS total_value
        FROM top_map
        WHERE pincode IS NOT NULL
        GROUP BY pincode
        ORDER BY total_value DESC
        LIMIT 10;
    """,

    "Spending Surge: Q4 vs Q1 (Yearly Diff)": """
        SELECT state, year, 
               SUM(CASE WHEN quarter = 1 THEN transaction_amount ELSE 0 END) AS Q1_total,
               SUM(CASE WHEN quarter = 4 THEN transaction_amount ELSE 0 END) AS Q4_total,
               (SUM(CASE WHEN quarter = 4 THEN transaction_amount ELSE 0 END) - 
                SUM(CASE WHEN quarter = 1 THEN transaction_amount ELSE 0 END)) AS surge
        FROM aggregated_transaction
        GROUP BY state, year
        HAVING Q1_total > 0 AND Q4_total > 0
        ORDER BY surge DESC
        LIMIT 10;
    """
}

# ✅ STEP 5: Execute queries and store results
from IPython.display import display, Markdown

# Loop through results and display each as a styled table
for title, query in sql_queries.items():
    try:
        df = pd.read_sql(query, engine)
        query_results[title] = df
        
        display(Markdown(f"### 📊 {title}"))
        display(df.head(20))  # Show top 20 rows for each insight

    except Exception as e:
        display(Markdown(f"### ❌ Error running `{title}`"))
        print(e)

### 📊 Device Brand Penetration (Top Brands)

brand    user_count
0    Xiaomi  1.130431e+10
1   Samsung  8.730848e+09
2      Vivo  8.130395e+09
3      Oppo  5.463253e+09
4    Others  3.678353e+09
5    Realme  2.859652e+09
6     Apple  1.247315e+09
7  Motorola  9.534295e+08
8   OnePlus  8.278037e+08
9    Huawei  7.426860e+08

### 📊 Top States by Transaction Consistency (Low Std Dev)

state       std_dev
0                         lakshadweep  2.779028e+07
1                             mizoram  7.352957e+08
2           andaman-&-nicobar-islands  1.169431e+09
3                              ladakh  1.602344e+09
4                              sikkim  1.830830e+09
5                            nagaland  2.118874e+09
6                           meghalaya  2.814819e+09
7                             manipur  2.952995e+09
8                             tripura  3.103964e+09
9  dadra-&-nagar-haveli-&-daman-&-diu  3.140983e+09

### 📊 Fastest Growing Insurance Districts (Q-over-Q Growth)

district  year  quarter  total_insurance
0     adilabad district  2020        2          73528.0
1     adilabad district  2020        3         430840.0
2     adilabad district  2020        4         560896.0
3     adilabad district  2021        1        1186120.0
4     adilabad district  2021        2         949528.0
5     adilabad district  2021        3        1766784.0
6     adilabad district  2021        4        1563616.0
7     adilabad district  2022        1        2509776.0
8     adilabad district  2022        2        2701848.0
9     adilabad district  2022        3        2679864.0
10    adilabad district  2022        4        3426104.0
11    adilabad district  2023        1        3431784.0
12    adilabad district  2023        2        2331432.0
13    adilabad district  2023        3        3762464.0
14    adilabad district  2023        4        3602288.0
15    adilabad district  2024        1        3508536.0
16    adilabad district  2024        2        2814760.0
17    adilabad district  2024        3        3773096.0
18    adilabad district  2024        4        3600256.0
19  agar malwa district  2020        2          21272.0

### 📊 Top Pin Codes by Transaction Value

pincode   total_value
0  500001  1.758589e+13
1  500034  1.666648e+13
2  560001  1.291180e+13
3  110006  1.134436e+13
4  411004  1.080457e+13
5  302016  9.168684e+12
6  800001  9.004022e+12
7  302012  6.995823e+12
8  462008  6.312524e+12
9  500072  6.156866e+12

### 📊 Spending Surge: Q4 vs Q1 (Yearly Diff)

state  year      Q1_total      Q4_total         surge
0      telangana  2022  2.004766e+13  3.573903e+13  1.569137e+13
1    maharashtra  2024  4.047479e+13  4.981234e+13  9.337556e+12
2      karnataka  2023  2.964273e+13  3.889652e+13  9.253791e+12
3      telangana  2023  2.847098e+13  3.739701e+13  8.926029e+12
4  uttar-pradesh  2024  2.973052e+13  3.856928e+13  8.838762e+12
5    maharashtra  2023  2.834743e+13  3.717618e+13  8.828749e+12
6    maharashtra  2021  9.238648e+12  1.770430e+13  8.465654e+12
7      karnataka  2024  4.182298e+13  4.982184e+13  7.998856e+12
8      telangana  2021  9.592311e+12  1.756510e+13  7.972787e+12
9      telangana  2024  3.844523e+13  4.630861e+13  7.863375e+12

**📝 Explanation:** This cell performs the following task:

 📊 1. Device Brand Penetration (Top Brands)
>What it reveals:
This insight shows the distribution of registered users by mobile device brand. Brands like Xiaomi, Samsung, and Vivo dominate the PhonePe user base.

> Why it matters:
Understanding device penetration helps in optimizing app performance and marketing strategies. For example, Xiaomi users might be more price-sensitive, guiding how to push new offers or features.

-----------------------------------------------------------------------------------------------------------------------------------------------------

📉 2. Top States by Transaction Consistency (Low Std Dev)
>What it reveals:
This query measures the standard deviation of transaction amounts over time per state. A lower standard deviation (e.g., in Lakshadweep or Mizoram) implies consistent transaction behavior.

>Why it matters:
Consistency suggests stable user engagement and economic patterns — ideal for testing new financial products or services.

-----------------------------------------------------------------------------------------------------------------------------------------------------

📈 3. Fastest Growing Insurance Districts (Q-over-Q Growth)
>What it reveals:
Districts like Adilabad show steadily increasing insurance transaction amounts quarter over quarter.

>Why it matters:
These are high-potential areas for expanding insurance offerings, user education, or targeted upselling.

-----------------------------------------------------------------------------------------------------------------------------------------------------

📌 4. Top Pin Codes by Transaction Value
>What it reveals:
This highlights pin codes (e.g., 500001, 500034) with the highest total transaction values.

>Why it matters:
These locations represent high-value zones — ideal for focused marketing, premium services, or merchant engagement.

-----------------------------------------------------------------------------------------------------------------------------------------------------

🔁 5. Spending Surge: Q4 vs Q1 (Yearly Diff)
>What it reveals:
Compares total spending in Q4 vs Q1 across states and years. States like Telangana and Maharashtra exhibit huge spikes in Q4.

>Why it matters:
This surge likely aligns with festive seasons or year-end sales, offering key timing for campaigns and service

-----------------------------------------------------------------------------------------------------------------------------------------------------

In [5]:
advanced_sql_queries = {
    "Brand Loyalty Index by State": """
        SELECT state, brand, brand_count,
               RANK() OVER (PARTITION BY state ORDER BY brand_count DESC) AS brand_rank,
               SUM(brand_count) OVER (PARTITION BY state) AS total_users,
               ROUND((brand_count / SUM(brand_count) OVER (PARTITION BY state)) * 100, 2) AS loyalty_percentage
        FROM aggregated_user
    """,

    "Inactive High-Value Districts": """
        SELECT m.state, m.district, m.amount, u.app_opens
        FROM map_map m
        JOIN map_user u ON m.state = u.state AND m.district = u.district AND m.year = u.year AND m.quarter = u.quarter
        WHERE u.app_opens < 1000 AND m.amount > 10000000
        ORDER BY m.amount DESC;
    """,

    "Insurance Penetration Efficiency": """
        SELECT state, SUM(amount) AS insurance_total, SUM(app_opens) AS app_opens,
               ROUND(SUM(amount)/NULLIF(SUM(app_opens), 0), 2) AS insurance_efficiency
        FROM map_insurance
        JOIN map_user USING(state, year, quarter, district)
        GROUP BY state
        HAVING insurance_efficiency IS NOT NULL
        ORDER BY insurance_efficiency DESC
        LIMIT 10;
    """,

    "Seasonal Transaction Spike Index (Q4 vs Q3)": """
        SELECT state, year,
               SUM(CASE WHEN quarter = 3 THEN transaction_amount ELSE 0 END) AS Q3_total,
               SUM(CASE WHEN quarter = 4 THEN transaction_amount ELSE 0 END) AS Q4_total,
               (SUM(CASE WHEN quarter = 4 THEN transaction_amount ELSE 0 END) - 
                SUM(CASE WHEN quarter = 3 THEN transaction_amount ELSE 0 END)) AS spike
        FROM aggregated_transaction
        GROUP BY state, year
        HAVING Q3_total > 0 AND Q4_total > 0
        ORDER BY spike DESC
        LIMIT 10;
    """,

    "Transaction Diversity Score by State": """
        SELECT state, COUNT(DISTINCT transaction_type) AS diversity_score
        FROM aggregated_transaction
        GROUP BY state
        ORDER BY diversity_score DESC
        LIMIT 10;
    
    """
}

# STEP 5: Execute and display tables
for title, query in advanced_sql_queries.items():
    try:
        df = pd.read_sql(query, engine)
        display(Markdown(f"### 📌 {title}"))
        display(df.head(20))
    except Exception as e:
        display(Markdown(f"### ❌ Error running `{title}`"))
        print(e)

### 📌 Brand Loyalty Index by State

state   brand  brand_count  brand_rank  total_users  \
0   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
1   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
2   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
3   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
4   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
5   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
6   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
7   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
8   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
9   andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
10  andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
11  andaman-&-nicobar-islands  Realme          760        1808    8104798.0   
12  andaman-&-nicobar-islands   Apple          735        1821    8104798.0   
13  andaman-&-nicobar-islands   Apple          735        1821    8104798.0   
14  andaman-&-nicobar-islands   Apple          735        1821    8104798.0   
15  andaman-&-nicobar-islands   Apple          735        1821    8104798.0   
16  andaman-&-nicobar-islands   Apple          735        1821    8104798.0   
17  andaman-&-nicobar-islands   Apple          735        1821    8104798.0   
18  andaman-&-nicobar-islands   Apple          735        1821    8104798.0   
19  andaman-&-nicobar-islands   Apple          735        1821    8104798.0   

    loyalty_percentage  
0                 0.01  
1                 0.01  
2                 0.01  
3                 0.01  
4                 0.01  
5                 0.01  
6                 0.01  
7                 0.01  
8                 0.01  
9                 0.01  
10                0.01  
11                0.01  
12                0.01  
13                0.01  
14                0.01  
15                0.01  
16                0.01  
17                0.01  
18                0.01  
19                0.01

### 📌 Inactive High-Value Districts

state                  district        amount  app_opens
0   karnataka  bengaluru urban district  8.211120e+10          0
1   karnataka  bengaluru urban district  8.211120e+10          0
2   karnataka  bengaluru urban district  8.211120e+10          0
3   karnataka  bengaluru urban district  8.211120e+10          0
4   karnataka  bengaluru urban district  8.211120e+10          0
5   karnataka  bengaluru urban district  8.211120e+10          0
6   karnataka  bengaluru urban district  8.211120e+10          0
7   karnataka  bengaluru urban district  8.211120e+10          0
8   karnataka  bengaluru urban district  8.211120e+10          0
9   karnataka  bengaluru urban district  8.211120e+10          0
10  karnataka  bengaluru urban district  8.211120e+10          0
11  karnataka  bengaluru urban district  8.211120e+10          0
12  karnataka  bengaluru urban district  8.211120e+10          0
13  karnataka  bengaluru urban district  8.211120e+10          0
14  karnataka  bengaluru urban district  8.211120e+10          0
15  karnataka  bengaluru urban district  8.211120e+10          0
16  karnataka  bengaluru urban district  8.211120e+10          0
17  karnataka  bengaluru urban district  8.211120e+10          0
18  karnataka  bengaluru urban district  8.211120e+10          0
19  karnataka  bengaluru urban district  8.211120e+10          0

### 📌 Insurance Penetration Efficiency

state  insurance_total     app_opens  \
0                              kerala     1.050974e+11  3.073664e+11   
1                          puducherry     2.969668e+09  1.341916e+10   
2                          chandigarh     2.662871e+09  1.249790e+10   
3                               delhi     6.522847e+10  3.431238e+11   
4                                 goa     6.698400e+09  4.070283e+10   
5                         lakshadweep     4.992296e+07  3.934910e+08   
6           andaman-&-nicobar-islands     1.550768e+09  1.355139e+10   
7                             haryana     6.647850e+10  6.251479e+11   
8                         west-bengal     8.419600e+10  8.213081e+11   
9  dadra-&-nagar-haveli-&-daman-&-diu     1.439393e+09  1.566259e+10   

   insurance_efficiency  
0                  0.34  
1                  0.22  
2                  0.21  
3                  0.19  
4                  0.16  
5                  0.13  
6                  0.11  
7                  0.11  
8                  0.10  
9                  0.09

### 📌 Seasonal Transaction Spike Index (Q4 vs Q3)

state  year      Q3_total      Q4_total         spike
0  uttar-pradesh  2024  3.260970e+13  3.856928e+13  5.959576e+12
1    maharashtra  2024  4.424722e+13  4.981234e+13  5.565120e+12
2  uttar-pradesh  2023  2.234098e+13  2.708929e+13  4.748314e+12
3    maharashtra  2023  3.302607e+13  3.717618e+13  4.150113e+12
4      rajasthan  2024  2.966709e+13  3.351659e+13  3.849499e+12
5          bihar  2024  2.136337e+13  2.518713e+13  3.823765e+12
6    maharashtra  2021  1.391109e+13  1.770430e+13  3.793215e+12
7      telangana  2024  4.261580e+13  4.630861e+13  3.692801e+12
8          bihar  2023  1.377784e+13  1.721683e+13  3.438990e+12
9      telangana  2021  1.413947e+13  1.756510e+13  3.425628e+12

### 📌 Transaction Diversity Score by State

state  diversity_score
0           andaman-&-nicobar-islands                5
1                      andhra-pradesh                5
2                   arunachal-pradesh                5
3                               assam                5
4                               bihar                5
5                          chandigarh                5
6                        chhattisgarh                5
7  dadra-&-nagar-haveli-&-daman-&-diu                5
8                               delhi                5
9                                 goa                5

**📝 Explanation:** This cell performs the following task:

📌 1. Brand Loyalty Index by State
>What it shows:
Ranks mobile brands within each state based on user count, and calculates their loyalty percentage (share of total users in the state).

>Why it matters:
You can identify dominant brands per state (e.g., Realme in Andaman) and tailor marketing or performance optimization efforts for that user base.

-----------------------------------------------------------------------------------------------------------------------------------------------------

📌 2. Inactive High-Value Districts
>What it shows:
Districts with very high transaction amounts but zero or very low app opens.

>Why it matters:
Highlights potential institutional or third-party activity — or app usage issues (e.g., reliance on merchants but not consumers). Bengaluru Urban is a notable outlier here.

-----------------------------------------------------------------------------------------------------------------------------------------------------

📌 3. Insurance Penetration Efficiency
>What it shows:
Insurance value transacted per app open. This is a proxy for efficiency — how effective user sessions are at generating insurance sales.

>Why it matters:
States like Kerala and Puducherry show better ROI on user engagement. Ideal for expanding insurance-related product offerings.


-----------------------------------------------------------------------------------------------------------------------------------------------------

📌 4. Seasonal Transaction Spike Index (Q4 vs Q3)
>What it shows:
Compares total transaction values between Q4 and Q3, showing the seasonal spike — typically driven by festivals or end-of-year spending.

>Why it matters:
Helps time promotional campaigns and infrastructure scaling. Uttar Pradesh and Maharashtra saw the largest Q4 spikes.


-----------------------------------------------------------------------------------------------------------------------------------------------------
📌 5. Transaction Diversity Score by State

>What it shows:
Counts how many distinct payment categories are used in each state.

>Why it matters:
High diversity = mature, digitally fluent users. If every state shows a score of 5 (as in your screenshot), that may indicate the same number of categories is used everywhere — worth further review to ensure granularity.


-----------------------------------------------------------------------------------------------------------------------------------------------------


In [ ]:
import streamlit as st
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sqlalchemy import create_engine

# --- App Configuration ---
st.set_page_config("📊 PhonePe - Insightful Dashboard", layout="wide")
st.title("📊 Data-Driven Analytics with PhonePe Transactions")
st.markdown("Exploring deep insights into user behavior, insurance, and transaction performance using real data.")

# --- Database Connection ---
@st.cache_resource
def get_engine():
    return create_engine("mysql+pymysql://root:vikasashi@localhost:3306/phonepe1")

engine = get_engine()

# --- Sidebar Filters ---
st.sidebar.header("🔍 Global Filters")
state = st.sidebar.text_input("Filter by State (optional)", "")
year = st.sidebar.selectbox("Filter by Year (optional)", ["", 2018, 2019, 2020, 2021, 2022, 2023])
quarter = st.sidebar.selectbox("Filter by Quarter (optional)", ["", 1, 2, 3, 4])

def add_filters(query, alias=""):
    conditions = []
    if state:
        conditions.append(f"{alias}state = '{state}'")
    if year:
        conditions.append(f"{alias}year = {year}")
    if quarter:
        conditions.append(f"{alias}quarter = {quarter}")
    if conditions:
        query += " WHERE " + " AND ".join(conditions)
    return query

# --- Tabs for Sections ---
tabs = st.tabs(["📈 Transaction Insights", "👥 User Behavior", "🛡️ Insurance Analysis", "📌 Pin Code Impact", "💡 Strategic Findings"])

# --- Tab 1: Transaction Insights ---
with tabs[0]:
    st.header("📈 Top Performing States by Transaction Volume")
    base_query = "SELECT state, SUM(transaction_amount) AS total_amount FROM aggregated_transaction GROUP BY state ORDER BY total_amount DESC LIMIT 10"
    df_tx = pd.read_sql(base_query, engine)
    st.dataframe(df_tx)

    fig, ax = plt.subplots(figsize=(10, 5))
    sns.barplot(data=df_tx, x="state", y="total_amount", ax=ax)
    ax.set_title("Top 10 States by Transaction Value")
    plt.xticks(rotation=45)
    st.pyplot(fig)

    st.subheader("📊 Seasonal Spike: Q4 vs Q1 by State")
    surge_query = """
    SELECT state, year,
        SUM(CASE WHEN quarter = 1 THEN transaction_amount ELSE 0 END) AS Q1,
        SUM(CASE WHEN quarter = 4 THEN transaction_amount ELSE 0 END) AS Q4,
        (SUM(CASE WHEN quarter = 4 THEN transaction_amount ELSE 0 END) - SUM(CASE WHEN quarter = 1 THEN transaction_amount ELSE 0 END)) AS surge
    FROM aggregated_transaction
    GROUP BY state, year
    HAVING Q1 > 0 AND Q4 > 0
    ORDER BY surge DESC
    LIMIT 10;
    """
    df_surge = pd.read_sql(surge_query, engine)
    st.dataframe(df_surge)

# --- Tab 2: User Behavior ---
with tabs[1]:
    st.header("👥 Device Brand Dominance (Brand Loyalty %)")
    loyalty_query = """
    SELECT state, brand, brand_count,
           ROUND((brand_count / total) * 100, 2) AS loyalty_pct
    FROM (
        SELECT state, brand, brand_count,
               SUM(brand_count) OVER (PARTITION BY state) AS total,
               ROW_NUMBER() OVER (PARTITION BY state ORDER BY brand_count DESC) AS rn
        FROM aggregated_user
    ) AS ranked
    WHERE rn = 1
    ORDER BY loyalty_pct DESC
    """
    df_loyalty = pd.read_sql(loyalty_query, engine)
    st.dataframe(df_loyalty)

    fig, ax = plt.subplots(figsize=(10, 5))
    sns.barplot(data=df_loyalty.head(10), x="state", y="loyalty_pct", hue="brand", ax=ax)
    ax.set_title("Top Brand Loyalty by State")
    plt.xticks(rotation=45)
    st.pyplot(fig)

    st.subheader("📉 States with Low App Usage (Drop-off Zones)")
    drop_query = """
    SELECT state, SUM(registered_users) AS total_users, SUM(app_opens) AS total_opens,
           ROUND(SUM(app_opens)/NULLIF(SUM(registered_users), 0), 2) AS usage_ratio
    FROM aggregated_user
    GROUP BY state
    HAVING usage_ratio < 0.5
    ORDER BY usage_ratio ASC
    LIMIT 10
    """
    df_drop = pd.read_sql(drop_query, engine)
    st.dataframe(df_drop)

# --- Tab 3: Insurance Analysis ---
with tabs[2]:
    st.header("🛡️ Insurance Penetration Efficiency by State")
    insurance_eff_query = """
    SELECT state, SUM(amount) AS total_insurance, SUM(app_opens) AS app_opens,
           ROUND(SUM(amount)/NULLIF(SUM(app_opens), 0), 2) AS efficiency
    FROM map_insurance
    JOIN map_user USING(state, year, quarter, district)
    GROUP BY state
    ORDER BY efficiency DESC
    LIMIT 10;
    """
    df_eff = pd.read_sql(insurance_eff_query, engine)
    st.dataframe(df_eff)

    fig, ax = plt.subplots(figsize=(10, 5))
    sns.barplot(data=df_eff, x="state", y="efficiency", ax=ax)
    ax.set_title("Insurance Amount per App Open (Efficiency Score)")
    plt.xticks(rotation=45)
    st.pyplot(fig)

# --- Tab 4: Pin Code Impact ---
with tabs[3]:
    st.header("📌 Top Pincodes by Transaction Value")
    pin_query = """
    SELECT pincode, SUM(amount) AS total_value
    FROM top_map
    WHERE pincode IS NOT NULL
    GROUP BY pincode
    ORDER BY total_value DESC
    LIMIT 10;
    """
    df_pin = pd.read_sql(pin_query, engine)
    st.dataframe(df_pin)

    fig, ax = plt.subplots(figsize=(10, 5))
    sns.barplot(data=df_pin, x="pincode", y="total_value", ax=ax)
    ax.set_title("Top 10 High-Value Transaction Pincodes")
    plt.xticks(rotation=45)
    st.pyplot(fig)

# --- Tab 5: Strategic Findings ---
with tabs[4]:
    st.header("💡 Market Diversity & Premium Pockets")

    diversity_query = """
    SELECT state, COUNT(DISTINCT transaction_type) AS variety
    FROM aggregated_transaction
    GROUP BY state
    ORDER BY variety DESC
    LIMIT 10;
    """
    df_variety = pd.read_sql(diversity_query, engine)
    st.dataframe(df_variety)

    fig, ax = plt.subplots(figsize=(10, 5))
    sns.barplot(data=df_variety, x="state", y="variety", ax=ax)
    ax.set_title("States with Most Diverse Payment Behaviors")
    plt.xticks(rotation=45)
    st.pyplot(fig)

    st.subheader("🔐 High Insurance with Low Users (Premium Pockets)")
    high_insurance_query = """
    SELECT i.state, i.district, SUM(i.amount) AS insurance_amt, SUM(u.registered_users) AS users,
           ROUND(SUM(i.amount)/NULLIF(SUM(u.registered_users), 0), 2) AS premium_index
    FROM map_insurance i
    JOIN map_user u ON i.state = u.state AND i.district = u.district AND i.year = u.year AND i.quarter = u.quarter
    GROUP BY i.state, i.district
    HAVING users < 5000 AND insurance_amt > 100000
    ORDER BY premium_index DESC
    LIMIT 10;
    """
    df_premium = pd.read_sql(high_insurance_query, engine)
    st.dataframe(df_premium)


# --- Additional Insights in Strategic Tab ---
    st.subheader("📊 Fastest Growing Insurance Districts Over Time")
    growth_query = """
    SELECT district, year, quarter, SUM(amount) AS total_insurance
    FROM map_insurance
    GROUP BY district, year, quarter
    ORDER BY district, year, quarter
    """
    df_growth = pd.read_sql(growth_query, engine)
    st.dataframe(df_growth.head(20))

    st.subheader("📊 States with High Transactions but Low Diversity (Concentration Risk)")
    concentration_query = """
    SELECT state, SUM(transaction_amount) AS total_amount,
           COUNT(DISTINCT transaction_type) AS type_count
    FROM aggregated_transaction
    GROUP BY state
    HAVING type_count <= 2
    ORDER BY total_amount DESC
    LIMIT 10;
    """
    df_concentration = pd.read_sql(concentration_query, engine)
    st.dataframe(df_concentration)

    st.subheader("📈 Top Districts with Growing App Engagement (Rising Opens)")
    app_growth_query = """
    SELECT district, year, quarter, SUM(app_opens) AS total_opens
    FROM map_user
    GROUP BY district, year, quarter
    HAVING total_opens > 50000
    ORDER BY total_opens DESC
    LIMIT 10;
    """
    df_app_growth = pd.read_sql(app_growth_query, engine)
    st.dataframe(df_app_growth)

    st.subheader("📊 States with High Insurance but Low Transaction Volume")
    ins_tx_query = """
    SELECT i.state, SUM(i.amount) AS insurance_total, SUM(t.transaction_amount) AS txn_total,
           ROUND(SUM(i.amount)/NULLIF(SUM(t.transaction_amount), 0), 2) AS insurance_ratio
    FROM map_insurance i
    JOIN aggregated_transaction t ON i.state = t.state AND i.year = t.year AND i.quarter = t.quarter
    GROUP BY i.state
    HAVING insurance_ratio > 0.5
    ORDER BY insurance_ratio DESC
    LIMIT 10;
    """
    df_ins_tx = pd.read_sql(ins_tx_query, engine)
    st.dataframe(df_ins_tx)

    st.subheader("📌 Most Active Districts Across All Categories")
    active_query = """
    SELECT mu.district,
       SUM(mu.registered_users) AS users,
       SUM(mu.app_opens) AS opens,
       SUM(mm.count) AS txns,
       SUM(mi.count) AS insurances
    FROM map_user mu
    JOIN map_map mm ON mu.state = mm.state AND mu.district = mm.district AND mu.year = mm.year AND mu.quarter = mm.quarter
    JOIN map_insurance mi ON mu.state = mi.state AND mu.district = mi.district AND mu.year = mi.year AND mu.quarter = mi.quarter
    GROUP BY mu.district
    ORDER BY txns DESC, opens DESC
    LIMIT 10;

    """
    df_active = pd.read_sql(active_query, engine)
    st.dataframe(df_active)


**📝 Explanation:** This cell performs the following task:

> This code Written is to be run in cmd, which will display the results on webpage furnished with visualizations with insights we have generated above.
